In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns

plt.style.use('fivethirtyeight')
%config InlineBackend.figure_format = 'svg'
%matplotlib inline

matplotlib.rcParams.update({'font.size': 10})
pd.options.display.max_columns = 30

**Описание датасета**

* **Id** - идентификационный номер квартиры
* **DistrictId** - идентификационный номер района
* **Rooms** - количество комнат
* **Square** - площадь
* **LifeSquare** - жилая площадь
* **KitchenSquare** - площадь кухни
* **Floor** - этаж
* **HouseFloor** - количество этажей в доме
* **HouseYear** - год постройки дома
* **Ecology_1, Ecology_2, Ecology_3** - экологические показатели местности
* **Social_1, Social_2, Social_3** - социальные показатели местности
* **Healthcare_1, Helthcare_2** - показатели местности, связанные с охраной здоровья
* **Shops_1, Shops_2** - показатели, связанные с наличием магазинов, торговых центров
* **Price** - цена квартиры

In [2]:
train = pd.read_csv('train.csv')
train.head(3)

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
0  14038          35    2.0  47.981561   29.442751            6.0      7   
1  15053          41    3.0  65.683640   40.049543            8.0      7   
2   4765          53    2.0  44.947953   29.197612            0.0      8   

   HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  \
0         9.0       1969   0.089040         B         B        33      7976   
1         9.0       1978   0.000070         B         B        46     10309   
2        12.0       1968   0.049637         B         B        34      7759   

   Social_3  Healthcare_1  Helthcare_2  Shops_1 Shops_2          Price  
0         5           NaN            0       11       B  184966.930730  
1         1         240.0            1       16       B  300009.450063  
2         0         229.0            1        3       B  220925.908524

In [3]:
test = pd.read_csv('test.csv')
test.head(3)

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
0    725          58    2.0  49.882643   33.432782            6.0      6   
1  15856          74    2.0  69.263183         NaN            1.0      6   
2   5480         190    1.0  13.597819   15.948246           12.0      2   

   HouseFloor  HouseYear  Ecology_1 Ecology_2 Ecology_3  Social_1  Social_2  \
0        14.0       1972   0.310199         B         B        11      2748   
1         1.0       1977   0.075779         B         B         6      1437   
2         5.0       1909   0.000000         B         B        30      7538   

   Social_3  Healthcare_1  Helthcare_2  Shops_1 Shops_2  
0         1           NaN            0        0       B  
1         3           NaN            0        2       B  
2        87        4702.0            5        5       B

In [4]:
train.shape

(10000, 20)

In [5]:
test.shape

(5000, 19)

#### Приведение типов данных

In [6]:
train.dtypes

Id                 int64
DistrictId         int64
Rooms            float64
Square           float64
LifeSquare       float64
KitchenSquare    float64
Floor              int64
HouseFloor       float64
HouseYear          int64
Ecology_1        float64
Ecology_2         object
Ecology_3         object
Social_1           int64
Social_2           int64
Social_3           int64
Healthcare_1     float64
Helthcare_2        int64
Shops_1            int64
Shops_2           object
Price            float64
dtype: object

In [7]:
test.dtypes

Id                 int64
DistrictId         int64
Rooms            float64
Square           float64
LifeSquare       float64
KitchenSquare    float64
Floor              int64
HouseFloor       float64
HouseYear          int64
Ecology_1        float64
Ecology_2         object
Ecology_3         object
Social_1           int64
Social_2           int64
Social_3           int64
Healthcare_1     float64
Helthcare_2        int64
Shops_1            int64
Shops_2           object
dtype: object

In [8]:
train['Id'].dtype

dtype('int64')

In [9]:
test['Id'].dtype

dtype('int64')

In [10]:
train['Id'] = train['Id'].astype(str)
train['Id'].dtype

dtype('O')

In [11]:
test['Id'] = test['Id'].astype(str)
test['Id'].dtype

dtype('O')

In [12]:
train.describe()

DistrictId         Rooms        Square   LifeSquare  KitchenSquare  \
count  10000.000000  10000.000000  10000.000000  7887.000000   10000.000000   
mean      50.400800      1.890500     56.315775    37.199645       6.273300   
std       43.587592      0.839512     21.058732    86.241209      28.560917   
min        0.000000      0.000000      1.136859     0.370619       0.000000   
25%       20.000000      1.000000     41.774881    22.769832       1.000000   
50%       36.000000      2.000000     52.513310    32.781260       6.000000   
75%       75.000000      2.000000     65.900625    45.128803       9.000000   
max      209.000000     19.000000    641.065193  7480.592129    2014.000000   

              Floor    HouseFloor     HouseYear     Ecology_1      Social_1  \
count  10000.000000  10000.000000  1.000000e+04  10000.000000  10000.000000   
mean       8.526700     12.609400  3.990166e+03      0.118858     24.687000   
std        5.241148      6.775974  2.005003e+05      0.119025     17.532614   
min        1.000000      0.000000  1.910000e+03      0.000000      0.000000   
25%        4.000000      9.000000  1.974000e+03      0.017647      6.000000   
50%        7.000000     13.000000  1.977000e+03      0.075424     25.000000   
75%       12.000000     17.000000  2.001000e+03      0.195781     36.000000   
max       42.000000    117.000000  2.005201e+07      0.521867     74.000000   

           Social_2      Social_3  Healthcare_1   Helthcare_2       Shops_1  \
count  10000.000000  10000.000000   5202.000000  10000.000000  10000.000000   
mean    5352.157400      8.039200   1142.904460      1.319500      4.231300   
std     4006.799803     23.831875   1021.517264      1.493601      4.806341   
min      168.000000      0.000000      0.000000      0.000000      0.000000   
25%     1564.000000      0.000000    350.000000      0.000000      1.000000   
50%     5285.000000      2.000000    900.000000      1.000000      3.000000   
75%     7227.000000      5.000000   1548.000000      2.000000      6.000000   
max    19083.000000    141.000000   4849.000000      6.000000     23.000000   

               Price  
count   10000.000000  
mean   214138.857399  
std     92872.293865  
min     59174.778028  
25%    153872.633942  
50%    192269.644879  
75%    249135.462171  
max    633233.466570

In [13]:
test.describe()

DistrictId        Rooms       Square   LifeSquare  KitchenSquare  \
count  5000.000000  5000.000000  5000.000000  3959.000000    5000.000000   
mean     51.279200     1.910000    56.449500    36.158810       5.976800   
std      44.179466     0.838594    19.092787    17.825287       9.950018   
min       0.000000     0.000000     1.378543     0.333490       0.000000   
25%      21.000000     1.000000    41.906231    23.092026       1.000000   
50%      37.000000     2.000000    52.921340    32.925087       6.000000   
75%      77.000000     2.000000    66.285129    45.174091       9.000000   
max     212.000000    17.000000   223.453689   303.071094     620.000000   

             Floor   HouseFloor    HouseYear    Ecology_1     Social_1  \
count  5000.000000  5000.000000  5000.000000  5000.000000  5000.000000   
mean      8.632000    12.601000  1984.392600     0.119874    24.933800   
std       5.483228     6.789213    18.573149     0.120070    17.532202   
min       1.000000     0.000000  1908.000000     0.000000     0.000000   
25%       4.000000     9.000000  1973.000000     0.019509     6.000000   
50%       7.000000    12.000000  1977.000000     0.072158    25.000000   
75%      12.000000    17.000000  2000.000000     0.195781    36.000000   
max      78.000000    99.000000  2020.000000     0.521867    74.000000   

           Social_2     Social_3  Healthcare_1  Helthcare_2      Shops_1  
count   5000.000000  5000.000000   2623.000000   5000.00000  5000.000000  
mean    5406.900000     8.262600   1146.657263      1.31940     4.242800  
std     4026.614773    23.863762   1044.744231      1.47994     4.777365  
min      168.000000     0.000000      0.000000      0.00000     0.000000  
25%     1564.000000     0.000000    325.000000      0.00000     1.000000  
50%     5285.000000     2.000000    900.000000      1.00000     3.000000  
75%     7287.000000     5.000000   1548.000000      2.00000     6.000000  
max    19083.000000   141.000000   4849.000000      6.00000    23.000000

Что не нравится и хотелось бы скорректировать:

1. приведение категориальных признаков Ecology_2, Ecology_3, Shops_2 в бинарные
* некорректные значения HouseYear в train
* недостаток данных Healthcare_1
* недостаток данных LifeSquare
* max KitchenSquare выглядит слишком большим в обоих датасетах
* max LifeSquare выглядит слишком большим в train
* min LifeSquare выглядит слишком малым в обоих датасетах

Что касается площадей - в моем понимании Square > KitchenSquare + LifeSquare. Исходя из этой логики в дальнейшем будем пытаться скорректировать значения в датасетах

#### 1. Приведение категориальных признаков Ecology_2, Ecology_3, Shops_2 в бинарные
Проверим количество возможных значений в категориальных признаках Ecology_2, Ecology_3, Shops_2 датасета train:

In [14]:
print(train['Ecology_2'].value_counts()) 
print(train['Ecology_3'].value_counts()) 
print(train['Shops_2'].value_counts())

B    9903
A      97
Name: Ecology_2, dtype: int64
B    9725
A     275
Name: Ecology_3, dtype: int64
B    9175
A     825
Name: Shops_2, dtype: int64


Т.к. каждый из категориальных признаков Ecology_2, Ecology_3, Shops_2 содержит только два варианта значений (A или B), то мы можем оставить только одно из них, которое будет соответствовать '1', а '0' будет соответствовать другому варианту. Заметим также, что в этих признаках нет пропусков.

In [15]:
train = pd.concat([train, 
                   pd.get_dummies(pd.DataFrame({'Ecology_2':train['Ecology_2'], 
                                                'Ecology_3':train['Ecology_3'], 
                                                'Shops_2':train['Shops_2']}), 
                                  prefix=['Ecol_2', 'Ecol_3', 'Shops_2'], 
                                  prefix_sep='_', 
                                  drop_first=True, 
                                  dtype='int64')
                  ], axis=1)
train.drop(['Ecology_2', 'Ecology_3', 'Shops_2'], axis=1, inplace=True)
train.head()

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
0  14038          35    2.0  47.981561   29.442751            6.0      7   
1  15053          41    3.0  65.683640   40.049543            8.0      7   
2   4765          53    2.0  44.947953   29.197612            0.0      8   
3   5809          58    2.0  53.352981   52.731512            9.0      8   
4  10783          99    1.0  39.649192   23.776169            7.0     11   

   HouseFloor  HouseYear  Ecology_1  Social_1  Social_2  Social_3  \
0         9.0       1969   0.089040        33      7976         5   
1         9.0       1978   0.000070        46     10309         1   
2        12.0       1968   0.049637        34      7759         0   
3        17.0       1977   0.437885        23      5735         3   
4        12.0       1976   0.012339        35      5776         1   

   Healthcare_1  Helthcare_2  Shops_1          Price  Ecol_2_B  Ecol_3_B  \
0           NaN            0       11  184966.930730         1         1   
1         240.0            1       16  300009.450063         1         1   
2         229.0            1        3  220925.908524         1         1   
3        1084.0            0        5  175616.227217         1         1   
4        2078.0            2        4  150226.531644         1         1   

   Shops_2_B  
0          1  
1          1  
2          1  
3          1  
4          1

In [16]:
train.shape

(10000, 20)

Аналогично проверим количество возможных значений в категориальных признаках Ecology_2, Ecology_3, Shops_2 датасета test, а также убедимся в отсутствии пропусков, и проведем те же манипуляции над датасетом для приведения названных признаков в бинарные:

In [17]:
print(test['Ecology_2'].value_counts()) 
print(test['Ecology_3'].value_counts()) 
print(test['Shops_2'].value_counts())

B    4952
A      48
Name: Ecology_2, dtype: int64
B    4851
A     149
Name: Ecology_3, dtype: int64
B    4588
A     412
Name: Shops_2, dtype: int64


In [18]:
test = pd.concat([test, 
                   pd.get_dummies(pd.DataFrame({'Ecology_2':test['Ecology_2'], 
                                                'Ecology_3':test['Ecology_3'], 
                                                'Shops_2':test['Shops_2']}), 
                                  prefix=['Ecol_2', 'Ecol_3', 'Shops_2'], 
                                  prefix_sep='_', 
                                  drop_first=True, 
                                  dtype='int64')
                  ], axis=1)
test.drop(['Ecology_2', 'Ecology_3', 'Shops_2'], axis=1, inplace=True)
test.head()

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
0    725          58    2.0  49.882643   33.432782            6.0      6   
1  15856          74    2.0  69.263183         NaN            1.0      6   
2   5480         190    1.0  13.597819   15.948246           12.0      2   
3  15664          47    2.0  73.046609   51.940842            9.0     22   
4  14275          27    1.0  47.527111   43.387569            1.0     17   

   HouseFloor  HouseYear  Ecology_1  Social_1  Social_2  Social_3  \
0        14.0       1972   0.310199        11      2748         1   
1         1.0       1977   0.075779         6      1437         3   
2         5.0       1909   0.000000        30      7538        87   
3        22.0       2007   0.101872        23      4583         3   
4        17.0       2017   0.072158         2       629         1   

   Healthcare_1  Helthcare_2  Shops_1  Ecol_2_B  Ecol_3_B  Shops_2_B  
0           NaN            0        0         1         1          1  
1           NaN            0        2         1         1          1  
2        4702.0            5        5         1         1          1  
3           NaN            3        3         1         1          1  
4           NaN            0        0         1         1          0

In [19]:
test.shape

(5000, 19)

#### 2. Исправим значение HouseYear в датасете train

In [20]:
train.loc[train['HouseYear'] > 2020]

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
1497  10814         109    1.0  37.265070   20.239714            9.0      9   
4189  11607         147    2.0  44.791836   28.360393            5.0      4   

      HouseFloor  HouseYear  Ecology_1  Social_1  Social_2  Social_3  \
1497        12.0   20052011   0.136330        30      6141        10   
4189         9.0       4968   0.319809        25      4756        16   

      Healthcare_1  Helthcare_2  Shops_1          Price  Ecol_2_B  Ecol_3_B  \
1497         262.0            3        6  254084.534396         1         1   
4189        2857.0            5        8  243028.603096         1         1   

      Shops_2_B  
1497          1  
4189          1

Из предыдущего вывода видно 2 выброса c DistrictId 109 и 147. Т.к. это всего два значения, они не должны оказать сильного влияния на обучение, но, тем не менее являются выбросами. Предположу замену HouseYear в первом случае на 2011, а во втором на 1968.

In [21]:
train.loc[(train['HouseYear'] == 20052011), 'HouseYear'] = 2011
train.loc[(train['HouseYear'] == 4968), 'HouseYear'] = 1968
pd.concat([train.loc[1497:1497,:], train.loc[4189:4189,:]], axis=0)

Id  DistrictId  Rooms     Square  LifeSquare  KitchenSquare  Floor  \
1497  10814         109    1.0  37.265070   20.239714            9.0      9   
4189  11607         147    2.0  44.791836   28.360393            5.0      4   

      HouseFloor  HouseYear  Ecology_1  Social_1  Social_2  Social_3  \
1497        12.0       2011   0.136330        30      6141        10   
4189         9.0       1968   0.319809        25      4756        16   

      Healthcare_1  Helthcare_2  Shops_1          Price  Ecol_2_B  Ecol_3_B  \
1497         262.0            3        6  254084.534396         1         1   
4189        2857.0            5        8  243028.603096         1         1   

      Shops_2_B  
1497          1  
4189          1

#### 3. Заполним недостающие данные Healthcare_1

In [22]:
# рассмотрим таблицу корреляций признака Healthcare_1
train.corr()[['Healthcare_1']]

Healthcare_1
DistrictId         0.304335
Rooms              0.042857
Square            -0.039748
LifeSquare        -0.026867
KitchenSquare      0.009472
Floor             -0.138294
HouseFloor        -0.143973
HouseYear         -0.205527
Ecology_1         -0.043547
Social_1           0.077133
Social_2           0.050458
Social_3           0.133877
Healthcare_1       1.000000
Helthcare_2        0.203177
Shops_1            0.057758
Price              0.138755
Ecol_2_B          -0.031724
Ecol_3_B          -0.114715
Shops_2_B         -0.031083

Исходя из предыдущей таблицы и сведений о признаках предположу, что признак Healthcare_1 зависит от (в порядке убывания):
 - DistrictId
 - Social_1
 - Social_2
 - Social_3
 - Helthcare_2

Для начала, я заполню все доступные значения Healthcare_1 опираясь на сведения по DistrictId, т.к. считаю, что для ID района должен больше соответсвовать конкретный Healthcare_1. 
Т.е. буду действовать в следующей последовательности:
1. составлю список индексов для Healthcare_1 == NaN, чтобы потом пробегать его в цикле;
2. пробегая список индексов (п.1) найду значения DistrictId для Healthcare_1 == NaN;
3. найду значения Healthcare_1 != NaN для известных DistrictId (п.2);
4. подставлю найденное Healthcare_1 (п.3) туда где Healthcare_1 == NaN (п.2) в соответствии с DistrictId.

In [23]:
# пишем функцию для поиска верных значений в train для train
for i in train[train.Healthcare_1.isna()].index.tolist():
    dist_id = train[train['Healthcare_1'].isna()].loc[i][1] # находим значение 'DistrictId', для которого Healthcare_1.isna
    try:
        hc1 = np.float64(train[train['Healthcare_1'].notna()].loc[train[train['Healthcare_1'].notna()]['DistrictId'] == dist_id, 'Healthcare_1'].mode().tolist()[0]) # находим значение 'Healthcare_1', по 'DistrictId'
    except IndexError:
        try:
            hc1 = np.float64(test[test['Healthcare_1'].notna()].loc[test[test['Healthcare_1'].notna()]['DistrictId'] == dist_id, 'Healthcare_1'].mode().tolist()[0]) # находим значение 'Healthcare_1', по 'DistrictId'
        except IndexError:
            hc1 = np.float64(None)
    train = train.copy()
    train.loc[i, 'Healthcare_1'] = hc1 # подставляем найденное 'Healthcare_1' вместо NaN
train.describe()

DistrictId         Rooms        Square   LifeSquare  KitchenSquare  \
count  10000.000000  10000.000000  10000.000000  7887.000000   10000.000000   
mean      50.400800      1.890500     56.315775    37.199645       6.273300   
std       43.587592      0.839512     21.058732    86.241209      28.560917   
min        0.000000      0.000000      1.136859     0.370619       0.000000   
25%       20.000000      1.000000     41.774881    22.769832       1.000000   
50%       36.000000      2.000000     52.513310    32.781260       6.000000   
75%       75.000000      2.000000     65.900625    45.128803       9.000000   
max      209.000000     19.000000    641.065193  7480.592129    2014.000000   

              Floor    HouseFloor     HouseYear     Ecology_1      Social_1  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean       8.526700     12.609400   1984.866300      0.118858     24.687000   
std        5.241148      6.775974     18.412271      0.119025     17.532614   
min        1.000000      0.000000   1910.000000      0.000000      0.000000   
25%        4.000000      9.000000   1974.000000      0.017647      6.000000   
50%        7.000000     13.000000   1977.000000      0.075424     25.000000   
75%       12.000000     17.000000   2001.000000      0.195781     36.000000   
max       42.000000    117.000000   2020.000000      0.521867     74.000000   

           Social_2      Social_3  Healthcare_1   Helthcare_2       Shops_1  \
count  10000.000000  10000.000000   8118.000000  10000.000000  10000.000000   
mean    5352.157400      8.039200   1043.768662      1.319500      4.231300   
std     4006.799803     23.831875   1014.192039      1.493601      4.806341   
min      168.000000      0.000000      0.000000      0.000000      0.000000   
25%     1564.000000      0.000000    240.000000      0.000000      1.000000   
50%     5285.000000      2.000000    750.000000      1.000000      3.000000   
75%     7227.000000      5.000000   1450.000000      2.000000      6.000000   
max    19083.000000    141.000000   4849.000000      6.000000     23.000000   

               Price      Ecol_2_B      Ecol_3_B     Shops_2_B  
count   10000.000000  10000.000000  10000.000000  10000.000000  
mean   214138.857399      0.990300      0.972500      0.917500  
std     92872.293865      0.098015      0.163543      0.275139  
min     59174.778028      0.000000      0.000000      0.000000  
25%    153872.633942      1.000000      1.000000      1.000000  
50%    192269.644879      1.000000      1.000000      1.000000  
75%    249135.462171      1.000000      1.000000      1.000000  
max    633233.466570      1.000000      1.000000      1.000000

Теперь для train.Healthcare_1 известны 8118 значений из 10000

Проделаем тоже самое для датасета test

In [24]:
# пишем функцию для поиска верных значений в test для test
for i in test[test.Healthcare_1.isna()].index.tolist():
    dist_id = test[test['Healthcare_1'].isna()].loc[i][1] # находим значение 'DistrictId', для которого Healthcare_1.isna
    try:
        hc1 = np.float64(test[test['Healthcare_1'].notna()].loc[test[test['Healthcare_1'].notna()]['DistrictId'] == dist_id, 'Healthcare_1'].mode().tolist()[0]) # находим значение 'Healthcare_1', по 'DistrictId'
    except IndexError:
        try:
            hc1 = np.float64(train[train['Healthcare_1'].notna()].loc[train[train['Healthcare_1'].notna()]['DistrictId'] == dist_id, 'Healthcare_1'].mode().tolist()[0]) # находим значение 'Healthcare_1', по 'DistrictId'
        except IndexError:
            hc1 = np.float64(None)
    test = test.copy()
    test.loc[i, 'Healthcare_1'] = hc1 # подставляем найденное 'Healthcare_1' вместо NaN
test.describe()

DistrictId        Rooms       Square   LifeSquare  KitchenSquare  \
count  5000.000000  5000.000000  5000.000000  3959.000000    5000.000000   
mean     51.279200     1.910000    56.449500    36.158810       5.976800   
std      44.179466     0.838594    19.092787    17.825287       9.950018   
min       0.000000     0.000000     1.378543     0.333490       0.000000   
25%      21.000000     1.000000    41.906231    23.092026       1.000000   
50%      37.000000     2.000000    52.921340    32.925087       6.000000   
75%      77.000000     2.000000    66.285129    45.174091       9.000000   
max     212.000000    17.000000   223.453689   303.071094     620.000000   

             Floor   HouseFloor    HouseYear    Ecology_1     Social_1  \
count  5000.000000  5000.000000  5000.000000  5000.000000  5000.000000   
mean      8.632000    12.601000  1984.392600     0.119874    24.933800   
std       5.483228     6.789213    18.573149     0.120070    17.532202   
min       1.000000     0.000000  1908.000000     0.000000     0.000000   
25%       4.000000     9.000000  1973.000000     0.019509     6.000000   
50%       7.000000    12.000000  1977.000000     0.072158    25.000000   
75%      12.000000    17.000000  2000.000000     0.195781    36.000000   
max      78.000000    99.000000  2020.000000     0.521867    74.000000   

           Social_2     Social_3  Healthcare_1  Helthcare_2      Shops_1  \
count   5000.000000  5000.000000   4089.000000   5000.00000  5000.000000   
mean    5406.900000     8.262600   1069.109807      1.31940     4.242800   
std     4026.614773    23.863762   1042.937316      1.47994     4.777365   
min      168.000000     0.000000      0.000000      0.00000     0.000000   
25%     1564.000000     0.000000    240.000000      0.00000     1.000000   
50%     5285.000000     2.000000    810.000000      1.00000     3.000000   
75%     7287.000000     5.000000   1547.000000      2.00000     6.000000   
max    19083.000000   141.000000   4849.000000      6.00000    23.000000   

          Ecol_2_B     Ecol_3_B    Shops_2_B  
count  5000.000000  5000.000000  5000.000000  
mean      0.990400     0.970200     0.917600  
std       0.097518     0.170052     0.275001  
min       0.000000     0.000000     0.000000  
25%       1.000000     1.000000     1.000000  
50%       1.000000     1.000000     1.000000  
75%       1.000000     1.000000     1.000000  
max       1.000000     1.000000     1.000000

Теперь для test.Healthcare_1 известны 4089 значений из 5000

In [ ]:
1. Проверить функции без `if-else`
2. Сделать функцию `def` для них
3. Заполнить недостающие Healthcare_1
4. Далее по списку...
5. Посмотреть KitchenSquare и LifeSquare на гистограммах или на сортировке по убыванию/возростанию на наличие выбросов

####### пишем функцию 
for i in ind1[:5]:
    if True:
        dist_id = a_hc1_0.loc[i][0] # находим значение 'DistrictId', для которого Healthcare_1.isna
        try:
            hc1 = np.float64(a_hc1_1.loc[a_hc1_1['DistrictId'] == dist_id, 'Healthcare_1'].mode().tolist()[0]) # находим значение 'Healthcare_1', по 'DistrictId'
        except IndexError:
            hc1 = np.float64(None)
        a_hc1_0 = a_hc1_0.copy()
        a_hc1_0.loc[i, 'Healthcare_1'] = hc1 # подставляем найденное 'Healthcare_1' вместо NaN
        # a_hc1_1['DistrictId' == dist_id, ]
    else:
        continue
a_hc1_0

In [ ]:
hc1 = np.float64(a_hc1_1.loc[a_hc1_1['DistrictId'] == 74, 'Healthcare_1'].mode().tolist()[0])
hc1

In [ ]:
hc1 = np.float64(None)
hc1

In [ ]:
test.sort_values(by='DistrictId', ascending=True).DistrictId.unique()

In [ ]:
test[test['DistrictId'] == 0]
test.loc[test['DistrictId'] == 0, 'Healthcare_1'].median()

In [ ]:
df = a_hc1_0.loc[0:9, :]
df

In [ ]:
hc1 = np.float64(a_hc1_1.loc[a_hc1_1['DistrictId'] == 1, ['Healthcare_1']].median())
df = df.copy()
df.loc[0, 'Healthcare_1'] = hc1
df
#df.loc.__setitem__((0, ('Healthcare_1')), 3)
#df['Healthcare_1'][0] = 1
# hc1[0] = hc1[0].astype('float64')
# type(hc1[0])
#df['Healthcare_1'][0] = 111
# df['Healthcare_1'].loc[0] = 3
# df['Healthcare_1'][0] = hc1

In [ ]:
? pd._setitem_with_indexer()

In [ ]:
df.loc[df['total_rooms'] < df['total_bedrooms'], 'total_bedrooms'] = df.loc[df['total_rooms'] < df['total_bedrooms'], 'total_rooms'].mean()

In [ ]:
train[['Healthcare_1', 'DistrictId', 'Social_1', 
       'Social_2', 'Social_3', 'Helthcare_2']].sort_values(by='DistrictId', ascending=True).head(20)

In [ ]:
a = train.loc[train['DistrictId'] == 23, ['Healthcare_1', 'DistrictId', 'Social_1', 
       'Social_2', 'Social_3', 'Helthcare_2']]
a.loc[a['Healthcare_1'].notna()]

In [ ]:
train.loc[train['DistrictId'] == 23, ['Healthcare_1', 'DistrictId', 'Social_1', 
       'Social_2', 'Social_3', 'Helthcare_2']][train['Healthcare_1'].isna()]

In [ ]:
train.DistrictId.value_counts()

In [ ]:
train[(train['Healthcare_1'] == 0)]

In [ ]:
# Такая табличка поможет понять нам, на сколько отличаются статистические данные (средние, СКО, медиана) 
# в тренировочном и тестовом датасетах.
stat_inf = pd.DataFrame({'mean_delta':(train.mean()-test.mean())/train.mean()*100,
                         'std_delta':(train.std()-test.std())/train.std()*100, 
                         'median_delta':(train.median()-test.median())/train.median()*100,
                         'mean2_delta':(train.mean()-test.mean())/test.mean()*100,
                         'std2_delta':(train.std()-test.std())/test.std()*100, 
                         'median2_delta':(train.median()-test.median())/test.median()*100})
stat_inf

Из предыдущей таблицы мы видим, что статистические данные (средние, СКО, медиана) в целом из обеих таблиц довольно близки. За исключением: 
* HouseYear
* KitchenSquare
* LifeSquare

In [ ]:
df.loc[df['total_rooms'] < df['total_bedrooms'], 'total_bedrooms'] = df.loc[df['total_rooms'] < df['total_bedrooms'], 'total_rooms']/(df['total_rooms']/df['total_bedrooms']).mean()

In [ ]:
train[(train['DistrictId'] == 109)].sort_values(by='HouseYear', ascending=False)

In [ ]:
train.isna().sum()

In [ ]:
train.corr()

In [ ]:
plt.figure(figsize = (11,12))

sns.set(font_scale=0.7)
sns.heatmap(train.corr(), annot=True, linewidths=.5, cmap='GnBu')

plt.title('Correlation matrix')
plt.show()

Что влияет на площадь Square:
* Rooms
* LifeSquare
* Price

In [ ]:
train[['Square', 'LifeSquare', 'KitchenSquare']][train['LifeSquare'] < 10]

In [ ]:
a_hc1_1[a_hc1_1.DistrictId == 1].DistrictId.value_counts()

In [ ]:
train[['DistrictId', 'Healthcare_1', 'Social_1']][(train['DistrictId']==23) & (train['Social_1']==0)]

In [ ]:
train[['Square', 'LifeSquare', 'KitchenSquare']]

In [ ]:
train[['Square', 'LifeSquare', 'KitchenSquare']][train['Square'] < train['LifeSquare'] + train['KitchenSquare']]

In [ ]:
train[train['KitchenSquare'] == 0]

In [ ]:
a = train[['DistrictId']][train['Healthcare_1'].isna()]
a.sort_values(by='DistrictId', ascending=True).DistrictId.unique()

In [ ]:
b = train[['DistrictId']][train['Healthcare_1'].notna()]
b.sort_values(by='DistrictId', ascending=True).DistrictId.unique()

In [ ]:
a.DistrictId.value_counts()

In [ ]:
b.DistrictId.value_counts()

In [ ]:
? pd.isna

In [ ]:
# долго висела, не посчитала
# sns.pairplot(train_num_features);

In [ ]:
train_num_features = train.select_dtypes(include=['float64', 'int64'])
train_num_features.hist(figsize=(11,20), bins=30, grid=False);